From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <spottedleaf@spottedleaf.dev>
Date: Tue, 22 Dec 2020 12:32:13 -0800
Subject: [PATCH] Apply more strict limitations to books

For reference is the vanilla limits, this must be checked
on update
- Max book pages: 50
- Max Length: 256

Currently, patch limits to 50 pages and 512 characters.

diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index dd2ba5d433d4aeef0ac5c0a81acb7f263104c10b..df01083f826463cce714eccd5317791c781f04ed 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -900,6 +900,36 @@ public class PlayerConnection implements PacketListenerPlayIn {
 
     @Override
     public void a(PacketPlayInBEdit packetplayinbedit) {
+        // Tuinity start - apply strict vanilla limits to books
+        ItemStack bookStack = packetplayinbedit.b();
+        if (!bookStack.isEmpty() && bookStack.getTag() != null) {
+            NBTTagList pages = bookStack.getTag().getList("pages", 8);
+            if (pages != null && !pages.isEmpty()) {
+                if (pages.size() > 50) {
+                    this.minecraftServer.scheduleOnMain(() -> this.disconnect("Book too large!"));
+                    return;
+                }
+                for (int i = 0, len = pages.size(); i < len; ++i) {
+                    NBTBase base = pages.get(i);
+                    if (!(base instanceof NBTTagString)) {
+                        this.minecraftServer.scheduleOnMain(() -> this.disconnect("Book too large!"));
+                        return;
+                    }
+                    String string = ((NBTTagString)base).asString();
+                    if (string.length() <= 256) {
+                        continue;
+                    }
+
+                    if (string.length() > (256*2)) {
+                        this.minecraftServer.scheduleOnMain(() -> this.disconnect("Book too large!"));
+                        return;
+                    }
+
+                    // Don't check the full length of the string, it's too expensive, could be exploited.
+                }
+            }
+        }
+        // Tuinity end - apply strict vanilla limits to books
         // Paper start
         ItemStack testStack = packetplayinbedit.b(); // TODO(Proximyst): Add obfhelper here
         if (!server.isPrimaryThread() && !testStack.isEmpty() && testStack.getTag() != null) {
